src/gz/bufread.rs

```diff
@@ -338,34 +476,45 @@ impl<'a, T: Read> Read for Buffer<'a, T> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         if self.buf_cur == self.buf_max {
             let len = self.reader.read(buf)?;
-            self.buf.extend_from_slice(&buf[..len]);
+            self.part.buf.extend_from_slice(&buf[..len]);
             Ok(len)
         } else {
-            let len = (&self.buf[self.buf_cur..self.buf_max]).read(buf)?;
+            let len = (&self.part.buf[self.buf_cur..self.buf_max]).read(buf)?;
             self.buf_cur += len;
             Ok(len)
         }
     }
 }
 
+impl<'a, T> Buffer<'a, T> where T: std::io::Read {
+    // If we manage to read all the bytes, we reset the buffer
+    fn read_once(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.read_exact(buf)?;
+        let rlen = buf.len();
+        self.crc.update(buf);
+        self.part.buf.truncate(0);
+        return Ok(rlen);
+    }
+}
+
 impl<R: BufRead> GzDecoder<R> {
     /// Creates a new decoder from the given reader, immediately parsing the
     /// gzip header.
     pub fn new(mut r: R) -> GzDecoder<R> {
-        let mut buf = Vec::with_capacity(10); // minimum header length
+        let mut part = GzHeaderPartial::new();
         let mut header = None;
 
         let result = {
-            let mut reader = Buffer::new(&mut buf, &mut r);
-            read_gz_header(&mut reader)
+            let mut reader = Buffer::new(&mut part, &mut r);
+            read_gz_header_part(&mut reader)
         };
 
         let state = match result {
-            Ok(hdr) => {
-                header = Some(hdr);
+            Ok(()) => {
+                header = Some(part.take_header());
```

Comment:
`read_gz_header_part` cannot return directly the `GzHeader` structure because it would move part of the `GzHeaderPartial` which is behind a mutable reference.

Thus, the need to break in 2 phases : 
1 return `Ok`
2 take ownership of the partial header fields